### PR TITLE
New module : Hi-C

### LINKED ISSUE #54: new module : HiC
I will implement the HiC workflow with input from Fidel.. It would be an independent workflow..

### PR DESCRIPTION
**Pull request by vivekbhr**
closes #54 

This workflow performs : 
 - mapping using BWA mem
 - building hicMatrix (hicExplorer)
- correcting matrix : diagnostics plot and correction (hicExplorer)
- calling TADs (hicExplorer)

Output Dir structure : 

```
.
├── FASTQ
│   ├── example_R1.fastq.gz -> ../fastq/example_R1.fastq.gz
│   └── example_R2.fastq.gz -> ../fastq/example_R2.fastq.gz
├── FastQC
│   ├── example_R1_fastqc.html
│   ├── example_R1_fastqc.zip
│   ├── example_R2_fastqc.html
│   ├── example_R2_fastqc.zip
│   └── logs
│       ├── FastQC.example_R1.log
│       └── FastQC.example_R2.log
├── BWA
│   ├── example_R12_rf.bam
│   ├── example_R1.bam
│   ├── example_R1.log
│   │── example_R2.bam
│   ├── example_R2.log
├── HiC_matrices
│   ├── example_corrected.h5
│   ├── example_m3_corrected.h5
│   ├── example_m3_diagnostic_plot.pdf
│   ├── example_m3.h5
│   ├── example_rf.h5
│   ├── logs
│   │   └── example.log
│   ├── QCplots
│  
├── HiC_matrices_corrected
│   ├── example_rf.corrected.h5
│   └── logs
│       └── example_rf.log
└── tads
    ├── example_corrected_min10_max40_step1.5_boundaries.bed
    ├── example_corrected_min10_max40_step1.5_boundaries.gff
    ├── example_corrected_min10_max40_step1.5_domains.bed
    ├── example_corrected_min10_max40_step1.5_score.bedgraph
    ├── example_corrected_min10_max40_step1.5_tad_score.bm
    ├── example_corrected_min10_max40_step1.5_zscore_matrix.h5
    ├── example_rf_boundaries.bed
    ├── example_rf_boundaries.gff
    ├── example_rf_domains.bed
    ├── example_rf_score.bedgraph
    ├── example_rf_tad_score.bm
    ├── example_rf_zscore_matrix.h5
    ├── logs
       └── example_findTADs.log
├── cluster_logs
├── config.yaml
├── DpnII.bed
├── Hi-C.log

```

### CODE REVIEW

File: workflows/HiC/HiC [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                        dest="downsample",
+                        metavar="INT",
+                        help="downsample the given number of reads randomly from of each FASTQ file (default: '%(default)s')",
+                        type=int,
+                        default=defaults["downsample"])
+
+    optional.add_argument("--trim",
+                        dest="trim",
+                        action="store_true",
+                        help="activate trimming (default: '%(default)s')",
```
[the remark this answers is not in the document]
Comment (vivekbhr) [reply workflows/HiC/HiC]: done

File: workflows/HiC/HiC [hunk 2]
```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/env python3
+
+__version__ = "0.7"
+
+__description__ = """
+DNA-mapping workflow v{version} - MPI-IE workflow for Hi-C analysis
```
[the remark this answers is not in the document]
Comment (vivekbhr) [reply workflows/HiC/HiC]: fixed

File: workflows/HiC/defaults.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+## directory with fastq files
+indir:
+## preconfigured target genomes (mm9,mm10,dm3,...) , see /path/to/snakemake_workflows/shared/organisms/
+## Value can be also path to your own genome config file!
+genome:
+## FASTQ file extension (default: ".fastq.gz")
+ext: .fastq.gz
+## paired-end read name extension (default: ['_R1', "_R2"])
+reads: [_R1, _R2]
+mapping_prg: BWA
```
Comment (fidelram) [inline workflows/HiC/defaults.yaml]: What if bowtie is used? The parameters no longer apply. For bowtie we need --reorder. I think we need to allow only one aligner.

File: workflows/HiC/HiC [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        prog=sys.argv[0],
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description=textwrap.dedent(__description__),
+        add_help=False
+    )
+
+    ## positional/required
+    parser.add_argument("genome",
+                        metavar="GENOME",
+                        help="Genome acronym of target organism (supported: 'dm3', 'dm6', 'hs37d5', 'mm9', 'mm10', 'SchizoSPombe_ASM294v2')")
```
[the remark this answers is not in the document]
Comment (vivekbhr) [reply workflows/HiC/HiC]: I added a new doc `setting_up` for read the docs to explain how to setup the workflow (including adding new organisms). We should remove hardcoded organism names from here in future

File: workflows/HiC/defaults.yaml (hunk 3 above)
Comment (vivekbhr) [reply workflows/HiC/defaults.yaml]: at this moment only one aligner is used (BWA) so this variable is useless and can be removed

**Comment by vivekbhr** [comment]
all right I am merging this one..

**Pull request merged**